File: BOJ/8075.py
import sys
input = sys.stdin.readline

class UF:
    def __init__(self, n):
        self.par = list(range(n))
        self.rk = [0] * n

    def find(self, x):
        if self.par[x] != x:
            self.par[x] = self.find(self.par[x])
        return self.par[x]

    def merge(self, x, y):
        x = self.find(x)
        y = self.find(y)
        if x == y:
            return
        if self.rk[x] > self.rk[y]:
            x, y = y, x
        self.par[x] = y
        if self.rk[x] == self.rk[y]:
            self.rk[y] += 1


def solve():
    k = int(input())
    length = list(map(int, input().split()))

    n = int(input())
    tokens = []
    while len(tokens) < 3:
        tokens += input().split()

    s0 = tokens[0]
    n  = int(tokens[1])
    s1 = tokens[2]

    m = {}
    cnt = 0

    m[('0', 0)] = cnt; cnt += 1
    m[('1', 0)] = cnt; cnt += 1

    v = [[], []]

    for idx, s in enumerate([s0, s1]):
        for c in s:

            if c == '0' or c == '1':
                v[idx].append(m[(c, 0)])

            else:
                L = length[ord(c) - ord('a')]
                for j in range(L):
                    key = (c, j)
                    if key not in m:
                        m[key] = cnt
                        cnt += 1
                    v[idx].append(m[key])
    
    if (len(v[0]) != len(v[1])):
        print(0)
        return

    uf = UF(cnt)

    for a, b in zip(v[0], v[1]):
        uf.merge(a, b)
    
    if (uf.find(0) == uf.find(1)):
        print(0)
        return

    ans = sum(1 for i in range(cnt) if uf.find(i) == i)
    ans -= 2
    if (ans < 0):
        print(0)
    else:
        print(2 ** ans)


def main():
    t = int(input())
    for _ in range(t):
        solve()


if __name__ == "__main__":
    main()

File: BOJ/8075.cpp
#include <bits/stdc++.h>
using namespace std;

int k;
map<pair<int, int>, int> m;
int len[26];
int cnt;

struct UF {
	vector<int> par, rk;

	UF(int _sz) {
		par.resize(_sz);
		rk.resize(_sz);
		for (int i = 0; i < _sz; ++i) par[i] = i;
	}
	int find(int x) {
		if (par[x] == x) return x;
		par[x] = find(par[x]);
		return par[x];
	}
	void merge(int x, int y) {
		x = find(x); y = find(y);
		if (x == y) return;
		if (rk[x] > rk[y]) swap(x, y);
		par[x] = y;
		if (rk[x] == rk[y]) ++rk[y];
	}
};

void solve() {
    cin >> k;
    for (int i = 0; i < k; ++i) cin >> len[i];
    int n; cin >> n;
    string s[2]; cin >> s[0] >> n >> s[1];
    vector<int> v[2];

    cnt = 0; m.clear();
    m[make_pair('0', 0)] = cnt++;
    m[make_pair('1', 0)] = cnt++;

    for (int i = 0; i < 2; ++i) {
        for (auto c: s[i]) {
            if (c == '0' || c == '1') v[i].push_back(m[make_pair(c, 0)]);
            else {
                for (int j = 0; j < len[c - 'a']; ++j) {
                    if (m.find(make_pair(c, j)) == m.end()) m[make_pair(c, j)] = cnt++;
                    v[i].push_back(m[make_pair(c, j)]);
                }
            }
        }
    }

    UF uf {cnt};
    for (int i = 0; i < v[0].size(); ++i) uf.merge(v[0][i], v[1][i]);

    long long ans = 0;
    for (int i = 0; i < cnt; ++i) if (uf.find(i) == i) ++ans;
    ans -= 2;
    cout << pow(2, ans) << '\n';

}

int main() {
    ios::sync_with_stdio(0); cin.tie(0);
    int t; cin >> t;
    while (t--) solve();
}